Improve parsing diagnostic for negative supertrait bounds

hdhoang opened the pull request:
closes #33418 

r? @estebank

petrochenkov commented on src/libsyntax/ast.rs:
```diff
@@ -241,12 +241,13 @@ pub const CRATE_NODE_ID: NodeId = NodeId::from_u32_const(0);
 pub enum TraitBoundModifier {
     None,
     Maybe,
+    ErroneousNegative,
```
Could you avoid adding impossible state to AST and report the error immediately in the parser, rather than in AST validation?

hdhoang replied:
I started out that way, but a parse error aborted parsing right away, thus I had to split the tests into 2 files. I'll try going back to that and see how rustfix applies.

hdhoang commented on src/librustc_passes/ast_validation.rs:
```diff
@@ -110,6 +110,31 @@ impl<'a> AstValidator<'a> {
+                            BytePos(sp.lo().0 - 1 - spaces.len() as u32));
+                        err.span_suggestion_with_applicability(span_to_remove,
+                            "remove this trait bound",
+                            "".into(), Applicability::MachineApplicable);
```
This is still too crude to be machine-applicable:

```rust
trait Tr3: !Sync + Copy {}
```

will be left as broken `Tr3 + Copy {}`. Does anyone have a suggestion? I will see if moving this back to the parser provides more context to fix such cases properly.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/475980590) ([raw log](https://api.travis-ci.org/v3/job/475980590/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0ec3f578:start=1546782125819050653,finish=1546782196239804965,duration=70420754312
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[01:00:27] .................................................................................................... 3500/5299
[01:00:31] .................................................................................................... 3600/5299
[01:00:34] ......................................ii............................................................ 3700/5299
[01:00:36] .........................................................i.......................................... 3800/5299
[01:00:37] .........................................F.......................................................... 3900/5299
[01:00:42] .................................................................................................... 4100/5299
[01:00:52] .................................................................................................... 4200/5299
[01:00:55] .................................................................................................... 4300/5299
[01:00:59] .................................................................................................... 4400/5299
[01:00:59] .................................................................................................... 4400/5299
[01:01:02] ........................................................i........................................... 4500/5299
[01:01:08] .................................................................................................... 4600/5299
[01:01:12] .................................................................................................... 4700/5299
[01:01:15] .................................................................................................... 4800/5299
[01:01:19] .................................................................................................... 4900/5299
[01:01:22] .................................................................................................... 5000/5299
[01:01:31] - LL | trait Tr4: !Sized + !Send + Copy + !Sync + !Unpin {} //~ ERROR negative trait bounds are not supported
[01:01:31] -    |                                  --^^^^^
[01:01:31] + LL | trait Tr4: !Sized + Copy + !Sync + !Unpin {} //~ ERROR negative trait bounds are not supported
[01:01:31] +    |                                  --^^^^^^
[01:01:31] 47    |                                  help: remove this trait bound
[01:01:31] 48 
[01:01:31] 
[01:01:31] 49 error: negative trait bounds are not supported
[01:01:31] 49 error: negative trait bounds are not supported
[01:01:31] -   --> $DIR/issue-33418.rs:6:44
[01:01:31] -    |
[01:01:31] - LL | trait Tr4: !Sized + !Send + Copy + !Sync + !Unpin {} //~ ERROR negative trait bounds are not supported
[01:01:31] -    |                                          --^^^^^^
[01:01:31] -    |                                          help: remove this trait bound
[01:01:31] - 
[01:01:31] - error: negative trait bounds are not supported
[01:01:31] 58   --> $DIR/issue-33418.rs:7:12
[01:01:31] 58   --> $DIR/issue-33418.rs:7:12
[01:01:31] 59    |
[01:01:31] 60 LL | trait Tr5: !Sized + !Send {} //~ ERROR negative trait bounds are not supported
[01:01:31] 70    |                   |
[01:01:31] 71    |                   help: remove this trait bound
[01:01:31] 72 
[01:01:31] - error: aborting due to 9 previous errors
---
[01:01:31] 
[01:01:31] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:495:22
[01:01:31] 
[01:01:31] 
[01:01:31] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/travis_time:end:2a9fb0c4:start=1546782205172551672,finish=1546785897059313420,duration=3691886761748
travis_time:start:03060326
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Sun Jan  6 14:44:57 UTC 2019
Sun, 06 Jan 2019 14:44:57 GMT
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

estebank commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5107,7 +5109,32 @@ impl<'a> Parser<'a> {
                         self.expect(&token::CloseDelim(token::Paren))?;
                     }
-                    let poly_trait = PolyTraitRef::new(lifetime_defs, path, lo.to(self.prev_span));
+                    let poly_span = lo.to(self.prev_span);
```
Looking at [the travis output](https://travis-ci.org/rust-lang/rust/jobs/475980590#L2085) it seems that you need to fix up the suggestion span:

```
20	LL | trait Tr3: !Sync + Unpin {} //~ ERROR negative trait bounds are not supported
-	   |            ^^^^^--
-	   |            |
-	   |            help: remove this trait bound
+	   |          --^^^^^
+	   |          |
+	   |          help: remove this trait bound
```

hdhoang replied:
my bad, I was leaving code and test quite out of sync. With this span, I was trying to leave the fixed code syntax-complete, thus I tried to remove the following `+` instead of the preceding `:`. In the general case, I imagine this will require much more info passed into this function.

estebank commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5107,7 +5109,32 @@ impl<'a> Parser<'a> {
+                                                    - 1 // for the `+` or `:`
+                                                    - spaces.len() as u32))
+                                    }
+                            }
```
Instead of looking back in the `source_map`, wouldn't it make more sense to modify `parse_ty_param` to pass in the `:` span to `parse_generic_bounds_common` (so that you can remove it if necessary) and also keep track of the prior `+` [later in this method (`!self.eat_plus()`)](https://github.com/rust-lang/rust/pull/57364/files#diff-da9d34ca1d0f6beee2838cf02e07345cR5149)? That way you will use either one of those two spans to create the suggestion span. Something along the lines of:

```
last_plus_sp = None;
if self.eat_plus() {
    last_plus_sp = Some(self.last_span);
}

/* ... */

let span_to_remove = last_plus_sp.unwrap_or(colon_sp).until(poly_span);
```

hdhoang replied:
thank you for your mentoring! I was quite ready to give up on rustfixing this. I'll need some time to grok and implement your method, and making more edge cases.

Since this is the parser, iteration time degraded big time compared to modifying the AST validation pass. Do you have any tip for reducing edit->test time, e.g. `incremental` in config.toml? I promise not to waste the project's CI budget like that.

estebank replied:
Sadly, hacking on `parser.rs` is just inherently slow. I run `./x.py test src/test/ui --stage1 --bless`, and that takes a while. Learning to use the profiler can be worth it, but for the most part I just switch to working on something else/fetch tea while `rustc` does its thing :-|

stokhos commented:
Ping from triage @hdhoang Have you been able to make any progress on this?

hdhoang commented:
thank you, I have a rough outline of the necessary flow inside the loop. I will describe most of it later while working it out locally.

hdhoang commented:
I have pushed a WIP flow for review.

Firstly, I have renamed the supertraits in tests to make talking about them easier.

On to the parser, to pass the colon span into the loop, I have added a `colon_span` argument to `parse_generic_bounds`/`parse_generic_bounds_common`, then passing `DUMMY_SP` at irrelevant callsites. The name isn't great, but hopefully using a dummy span elsewhere makes it clear.

Next, I have implemented your `last_plus_span` flow, and created a flag to track whether we need to promote a plus to a colon to keep the code valid. I temporarily used `struct_span_err` here, but I think a warning/note/help message hanging on to `last_plus_span` is more appropriate. I will find a more suitable macro for that, and a clearer message for "will become the first one".

estebank commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5416,7 +5453,7 @@ impl<'a> Parser<'a> {
                 let ty = self.parse_ty()?;
                 if self.eat(&token::Colon) {
-                    let bounds = self.parse_generic_bounds()?;
+                    let bounds = self.parse_generic_bounds(syntax_pos::DUMMY_SP)?;
```
This makes me thing that either the argument should be an `Option<Span>`.

hdhoang commented:
thanks! I intend to update this weekend, and see how it goes

hdhoang commented:
hopefully I have addressed your review correctly. I have also downgraded the colon suggestion to a warning, since it's due to rustfix, not the original code

hdhoang commented:
ping @estebank, have you had a chance chance to look at this?

estebank commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5114,7 +5120,39 @@ impl<'a> Parser<'a> {
+                                ":".into(),
+                                Applicability::MachineApplicable)
+                            .emit();
+                    }
```
As neat as the output is, it's unadvisable to have multiple suggestions that need to be applied together for them to be correct in separate diagnostics. A more appropriate solution would be to batch all the errors to be emitted (or at least the elements needed to create diagnostics) and emit a single, unambiguous diagnostic with the full suggestion. This would have the nice side effect of also dealing with `Tr5` with a single diagnostic, instead of one per negative trait bound.

Do you feel this is reasonable?

estebank commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5114,7 +5120,39 @@ impl<'a> Parser<'a> {
+                                "".into(),
+                                Applicability::MachineApplicable);
+                        } else {
+                            err.help("remove this trait bound");
```
If the other comment is implemented, this is the branch where you would have to account for the trailing `+` span instead, and suggest the removal of it, instead of in the warning.

estebank commented on src/test/ui/parser/issue-33418.stderr:
```diff
@@ -0,0 +1,74 @@
+LL |     + !SuperB {} //~ ERROR negative trait bounds are not supported
+   |     --^^^^^^^
+   |     |
+   |     help: remove this trait bound
```
If the changes are implemented, this output would be:

```
error: negative trait bounds are not supported
  --> $DIR/issue-33418.rs:11:12
   |
LL | trait Tr5: !SuperA
   |            ^^^^^^^
LL |     + !SuperB {}
   |       ^^^^^^^
help: remove the trait bound:
   |
LL | trait Tr5 {}
   |         --
```

hdhoang commented on src/test/ui/parser/issue-33418.stderr:
```diff
@@ -0,0 +1,64 @@
+help: remove the trait bound
+   |
+LL | trait Tr3 + SuperB {} //~ ERROR negative trait bounds are not supported
+   |         --
```
These help messages come out nonsensical, but the code suggestions apply correctly. Is this okay?

hdhoang commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5350,17 +5350,25 @@ impl<'a> Parser<'a> {
+        let mut has_removed_colon = false;
+        let mut negative_bounds = Vec::new();
+        let mut last_plus_span = None;
+        let mut promoted_plus_span = None;
```
```suggestion
        let mut new_colon_span = None;
```

This should read better

hdhoang commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5379,7 +5387,19 @@ impl<'a> Parser<'a> {
+                            }).unwrap().to(poly_span));
+                    } else if has_removed_colon {
+                        has_removed_colon = false;
+                        promoted_plus_span = last_plus_span;
```
```suggestion
                        new_colon_span = last_plus_span;
```

hdhoang commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5393,14 +5413,33 @@ impl<'a> Parser<'a> {
+        if !negative_bounds.is_empty() {
+            let mut err = self.struct_span_err(negative_bounds.clone(),
+                                               "negative trait bounds are not supported");
+            if promoted_plus_span.is_some() {
```
```suggestion
            if new_colon_span.is_some() {
```

hdhoang replied:
```suggestion
            if new_colon_span.is_some() {
```

hdhoang commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5393,14 +5413,33 @@ impl<'a> Parser<'a> {
+            let mut err = self.struct_span_err(negative_bounds.clone(),
+                                               "negative trait bounds are not supported");
+            if promoted_plus_span.is_some() {
+                err.span_suggestion_short(promoted_plus_span.unwrap(),
```
```suggestion
                err.span_suggestion_short(new_colon_span.unwrap(),
```

hdhoang commented:
I have reconstructed a new bound list from the parsed `bounds`, unfortunately this will also delete linebreaks and compiletest`s comment. However I hope this method have minimal impact on the happy path.

hdhoang commented:
hi @estebank, how do you feel about the current approach?

estebank commented:
@hdhoang I've been away for the past week. I'll try to review today.

bors commented:
:umbrella: The latest upstream changes (presumably #58341) made this pull request unmergeable. Please resolve the merge conflicts.

hdhoang commented:
I have resolved the conflict locally, but still would like to wait to rebase & make use of #58296 before pushing

hdhoang commented:
Using `span_suggestion_hidden` gave the same output as `span_suggestion_short`, so I have skipped out on that change and only fixed the conflict.

estebank commented:
@bors r+

bors commented:
:pushpin: Commit a55f1c3da60797b5c705ef34433a1d3414f99bf6 has been approved by `estebank`

<!-- @bors r=estebank a55f1c3da60797b5c705ef34433a1d3414f99bf6 -->

bors commented:
:hourglass: Testing commit a55f1c3da60797b5c705ef34433a1d3414f99bf6 with merge 229d6c9e1f46fd011711f6f0a11a5c662467c74e...

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/101907766)

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/179763914) ([raw log](https://api.travis-ci.com/v3/job/179763914/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Attempting to download s3://rust-lang-ci-sccache2/docker/3c36f6ef4a6e74d796913872a6a5b69acaf0c4c6f133d83c18c6b8c8512a86632d37a5f770883d03ef722cf297828328256fc443de9aaf0d1ee873afc85e0786
[00:01:15] Attempting with retry: curl -f -L -C - -o /tmp/rustci_docker_cache https://s3-us-west-1.amazonaws.com/rust-lang-ci-sccache2/docker/3c36f6ef4a6e74d796913872a6a5b69acaf0c4c6f133d83c18c6b8c8512a86632d37a5f770883d03ef722cf297828328256fc443de9aaf0d1ee873afc85e0786
[00:01:15]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[00:01:15]                                  Dload  Upload   Total   Spent    Left  Speed
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

hdhoang commented:
Looks to be a network failure, please retry.

ljedrz commented:
@bors retry

bors commented:
@ljedrz: :key: Insufficient privileges: not in try users

ljedrz commented:
Oops, it seems that my try privileges haven't been deployed yet.

kennytm commented:
@bors retry

bors commented:
:umbrella: The latest upstream changes (presumably #58644) made this pull request unmergeable. Please resolve the merge conflicts.

hdhoang commented:
I'll get on this

hdhoang commented:
hi, this is good to go again. please ask @bors to retry.

ljedrz commented:
Ok, let's try again ^^.

@bors retry

bors commented:
@ljedrz: :key: Insufficient privileges: not in try users

kennytm commented:
After push to a PR it can't simply be retried, it must be r+'ed again.

bors commented:
:pushpin: Commit 7cfddfb4e40584dc206cedef7c70b5fecb03d6a1 has been approved by `estebank`

<!-- @bors r=estebank 7cfddfb4e40584dc206cedef7c70b5fecb03d6a1 -->

hdhoang commented:
thanks to everyone for guidance!
